Share stateless fake call rules

blairconrad opened the pull request:
Connects to #1489.

Avoids creation of a number of rule objects with every Fake, at the cost of a slightly more complicated implementation.

|                                            Method |                  Job |     Mean |     Error |    StdDev | Ratio | RatioSD |
|-------------------------------------------------- |--------------------- |---------:|----------:|----------:|------:|--------:|
| &#39;A.Fake&lt;SimpleClass&gt;(o =&gt; o.Implements&lt;IFake&gt;())&#39; |      4.10.0-alpha070 | 5.741 us | 0.6874 us | 0.4546 us |  1.00 |    0.00 |
| &#39;A.Fake&lt;SimpleClass&gt;(o =&gt; o.Implements&lt;IFake&gt;())&#39; | 4.9.0-share-rules001 | 5.208 us | 0.3416 us | 0.2260 us |  0.91 |    0.09 |
|                             A.Fake&lt;SimpleClass&gt;() |      4.10.0-alpha070 | 4.833 us | 0.3176 us | 0.1890 us |  1.00 |    0.00 |
|                             A.Fake&lt;SimpleClass&gt;() | 4.9.0-share-rules001 | 4.190 us | 0.1527 us | 0.0909 us |  0.87 |    0.05 |
|                         A.Fake&lt;LessSimpleClass&gt;() |      4.10.0-alpha070 | 9.777 us | 0.0985 us | 0.0586 us |  1.00 |    0.00 |
|                         A.Fake&lt;LessSimpleClass&gt;() | 4.9.0-share-rules001 | 9.733 us | 0.0836 us | 0.0437 us |  1.00 |    0.01 |
|                                   A.Fake&lt;IFake&gt;() |      4.10.0-alpha070 | 5.158 us | 0.5939 us | 0.3928 us |  1.00 |    0.00 |
|                                   A.Fake&lt;IFake&gt;() | 4.9.0-share-rules001 | 4.725 us | 0.1679 us | 0.1111 us |  0.92 |    0.07 |

![image](https://user-images.githubusercontent.com/3275797/48220734-1b584200-e35e-11e8-9c20-a828b889f06d.png)

blairconrad commented:
The first couple of commits were separated out to make the diffs easier. If you prefer, we could squash later. I'm not fussy.

blairconrad commented:
Um. Hold tight. I think I accidentally reordered some methods, making the diff hard.

blairconrad commented:
Okay. Unreordered the `ObjectMemberRule` methods. Review away!

blairconrad commented on tests/FakeItEasy.Specs/UnconfiguredFakeSpecs.cs:
```diff
@@ -141,6 +141,35 @@ public interface IFoo
+                .x(() => fake1 = A.Fake<FakedClass>());
+
+            "And a second faked class instance of the same type"
+                .x(() => fake1 = A.Fake<FakedClass>());
```
copy/paste error! Should be `fake2`. Will update sometime this evening.

blairconrad commented on tests/FakeItEasy.Tests/Core/FakeManagerTests.cs:
```diff
@@ -342,62 +339,14 @@ public void Should_return_hash_code_of_fake_manager_when_GetHashCode_has_not_bee
-            // Assert
-            A.CallTo(() => interceptedCall.SetReturnValue(true)).MustHaveHappened();
-        }
-
```
These started failing, because we don't use the faked `FakeManagerAccessor`. Moved to specs (which you saw above), as near as I was able.

blairconrad commented:
I added a late-breaking commit that reduces the overhead of the `EventRule`. Would eventually squash.

blairconrad commented on src/FakeItEasy/Core/FakeManager.cs:
```diff
@@ -241,5 +267,22 @@ private void UpdateLastSequenceNumber(int sequenceNumber)
+        /// it'll be supplied in the <see cref="ApplyBestRule(IInterceptedFakeObjectCall, out IFakeObjectCallRule)"/> method.
+        /// The only reason these rules extend <see cref="IFakeObjectCallRule"/> is that they are emitted through
+        /// <see cref="IInterceptionListener.OnAfterCallIntercepted"/>.
+        /// </summary>
```
As I hint at, this hack wouldn't be necessary if `IFakeObjectCallRule` didn't expose `Apply`. But it needs to in order for people to create their own rules.

But I don't think that `IInterceptionListener.OnAfterCallIntercepted` needs to be passed an object that has an `Apply` method; there's no need for `OnAfterCallIntercepted` to call `Apply`.
I don't even know why it would need to call `IsApplicable` or `NumberOfTimesToCall`. Shall we explore  changing this in a future major version? We could supply `OnAfterCallIntercepted` with a different interface that's a restricted version of `IFakeObjectCallRule`, or even remove the `IFakeObjectCallRule` parameter.

thomaslevesque replied:
Seems like a good idea. I don't see any value in having the call rule available in `OnAfterCallIntercepted`, except for information purposes. Maybe we could just pass a string describing the rule, or just the type of the rule.

thomaslevesque commented on src/FakeItEasy/Core/FakeManager.cs:
```diff
@@ -143,14 +141,10 @@ void IFakeCallProcessor.Process(IInterceptedFakeObjectCall fakeObjectCall)
             try
             {
-                ApplyRule(ruleToUse, fakeObjectCall);
+                this.ApplyBestRule(fakeObjectCall, out ruleUsed);
```
Is there a reason to return `ruleUsed` as an out parameter here? Can't it just be returned directly, since there's no other return value?

blairconrad replied:
It was a way to set the `ruleUsed` (and ensure that the caller knew what it was) before attempting the application of the rule. If the `.Apply` throws, we won't get an opportunity to return the rule. I could wrap each `Apply` call in `ApplyBestRule` with try/catches and return the rule regardless.

thomaslevesque replied:
Ah, I see. You can leave it as is, then.

blairconrad replied:
Sorry, I was unclear. In progress, I'm replacing `ApplyBestRule` with 

```
private IFakeObjectCallRule ChooseBestRule(IInterceptedFakeObjectCall fakeObjectCall)
{
    foreach (var preUserRule in this.preUserRules)
    {
        if (preUserRule.IsApplicableTo(fakeObjectCall))
        {
            return preUserRule;
        }
    }

    foreach (var rule in this.AllUserRules)
    {
        if (rule.Rule.IsApplicableTo(fakeObjectCall) && rule.HasNotBeenCalledSpecifiedNumberOfTimes())
        {
            rule.CalledNumberOfTimes++;
            return rule.Rule;
        }
    }

    foreach (var postUserRule in PostUserRules)
    {
        if (postUserRule.IsApplicableTo(fakeObjectCall))
        {
            return postUserRule;
        }
    }

    // The last rule in the postUserRules is applicable to all calls, so this line is here to satisfy
    // the compiler. We will always have returned by now.
    return null;
}
```

And then we can just call `Apply` right after. I think it's better than the `out`. But if you prefer the `out`, I can leave.

thomaslevesque replied:
Ah, I see. What bothers me with this approach is that `CalledNumberOfTimes` is incremented in the method that chooses the rule, rather than in the method that applies it...

blairconrad replied:
A fair point. Back to `out` with an explanatory comment?

thomaslevesque replied:
Sure, a comment wouldn't hurt!

blairconrad commented:
Thanks, @thomaslevesque. Good comments! Some of the requested changes break unit tests, so there will be a slight delay as I address that, but everything you ask about is achievable.

blairconrad commented:
Thanks, @thomaslevesque. Updated. I added two fixup commits, but the second one sort of spans two original commits. If ever you're happy with the changes, I'd be content to squash down to 1, unless you prefer some separation.

thomaslevesque commented on tests/FakeItEasy.Specs/UnconfiguredFakeSpecs.cs:
```diff
@@ -166,10 +179,26 @@ public static void EqualsDifferentInstanceOfSameFakedType(FakedClass fake1, Fake
+
+            "Then the hash codes are the same"
+                .x(() => hashCode.Should().Be(managerHashCode));
+        }
```
Isn't the fact that the fake and its manager have the same hashcode an implementation detail? i.e. do we really need a spec for it?

blairconrad replied:
You're right. I always struggle with how to test `GetHashCode`. We can make sure that fakes that are equal have the same hash code, but since we've a 1:1 map between fakes and their managers, the implemented behaviour devolves to "use reference semantics for `Equals` and `GetHashCode`". Then the only meaningful test I can think of is "a Fake generates the same hash code twice". Which isn't bad, but it is pretty weak.
I'd considered checking that two fakes give different hash codes, but that's not a requirement and the test could fail even when everything is working properly.

I can certainly add a test to ensure that the same fake's `GetHashCode` returns the same value more than once. And I can move the test back (and update it so it passes) if you like,  but the fact that we return the manager's hash code is meaningless enough that I'd almost rather remove the test altogether.

Have you a preference?

blairconrad replied:
Sorry, "remove the test that compares against the FakeManager's `GetHashCode` value".

blairconrad replied:
For now, "improved" the spec at least.

thomaslevesque replied:
> Have you a preference?

I think I'd remove this spec altogether.

thomaslevesque commented:
> If ever you're happy with the changes, I'd be content to squash down to 1, unless you prefer some separation.

At this point it's probably simpler to just squash everything.

blairconrad commented:
Thanks, @thomaslevesque. I appreciate the comments. The `GetHashCode` and "idempotent" ones were gold.
Squashed and pushed. I threw https://github.com/blairconrad/FakeItEasy/tree/share-rules-unsquashed up in case you wanted to see that the only change was to remove the spec(s).

thomaslevesque commented:
Thanks @blairconrad!

blairconrad commented:
Ha. I gave you so much misery over the bad changes, I'm not sure I deserve thanks. :smile:
